src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php

```diff
@@ -116,20 +122,25 @@ public function __construct(array $configuration, array $connectionCredentials =
         $this->claimInterval = $configuration['claim_interval'] ?? self::DEFAULT_OPTIONS['claim_interval'];
     }
 
-    public static function fromDsn(string $dsn, array $redisOptions = [], \Redis $redis = null): self
+    /**
+     * @param \Redis|\RedisCluster|null $redis
+     */
+    public static function fromDsn(string $dsn, array $redisOptions = [], $redis = null): self
     {
-        $url = $dsn;
+        if (false === strpos($dsn, ',')) {
+            $parsedUrl = self::parseDsn($dsn, $redisOptions);
+        } else {
+            $parsedUrls = array_map(function ($dsn) use (&$redisOptions) {
+                return self::parseDsn($dsn, $redisOptions);
+            }, explode(',', $dsn));
 
-        if (preg_match('#^redis:///([^:@])+$#', $dsn)) {
-            $url = str_replace('redis:', 'file:', $dsn);
-        }
+            // Merge all the URLs, the last one overrides the previous ones
+            $parsedUrl = array_merge(...$parsedUrls);
 
-        if (false === $parsedUrl = parse_url($url)) {
-            throw new InvalidArgumentException(sprintf('The given Redis DSN "%s" is invalid.', $dsn));
-        }
-        if (isset($parsedUrl['query'])) {
-            parse_str($parsedUrl['query'], $dsnOptions);
-            $redisOptions = array_merge($redisOptions, $dsnOptions);
+            // Regroup all the hosts in an array interpretable by RedisCluster
+            $parsedUrl['host'] = array_map(function ($parsedUrl) {
+                return $parsedUrl['host'].':'.($parsedUrl['port'] ?? 6379);
```

Comment:
this won't report things cleanly if there is no `$parsedUrl['host']`

Reply:
I've thought about it and, actually, I don't really know how to handle this. What should happen if `host`is missing? Should we throw?

Reply:
Now I'm thowing an exception when the host is missing when using Redis Cluster:

https://github.com/symfony/symfony/blob/35fa4917334ae7075b0048f7557619635ba01346/src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php#L148-L155